bootstrap: Build with `use-lld = true` fails under MSYS2 due to some path escaping issues

The build fails soon after start with a log like this:
```
we@WIN-GQ6AS7OLS34 MINGW64 ~/rust
$ ./x t -j1
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
Building stage0 tool tidy (x86_64-pc-windows-gnu)
   Compiling serde_derive v1.0.217
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,C:\\msys64\\tmp\\rustcIGBlzu\\list.def" "-fno-use-linker-plugin" "-Wl,--dynamicbase" "-Wl,--disable-auto-image-base" "-m64" "-Wl,--high-entropy-va" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "C:\\msys64\\tmp\\rustcIGBlzu\\symbols
.o" "<16 object files omitted>" "<sysroot>-bootstrap-tools\\release\\deps\\serde_derive-289e03abbb61a8ac.7h71bf3kbtp6y7yvtao8biynu.rcgu.rmeta" "<1 object files omitted>" "-Wl,-Bstatic" "C:\\msys64\\home\\we\\rust\\build\\x86_64-pc-windows-gnu\\stage0-bootstrap-tools\\rel
ease\\deps/{libsyn-c238c3c558a6e659.rlib,libquote-9a71ca2559253843.rlib,libproc_macro2-eff81bcdd40c9193.rlib,libunicode_ident-203f2b90658d2aa5.rlib}.rlib" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib/{libproc_macro-*,libstd-*,libpanic_unwind-*,libobject-*,libmemc
hr-*,libaddr2line-*,libgimli-*,libwindows_targets-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,liballoc-*,librustc_std_workspace_core-*,libcore-*,libcompiler_builtins-*}
.rlib" "-Wl,-Bdynamic" "-lkernel32" "-lkernel32" "-ladvapi32" "-lntdll" "-luserenv" "-lws2_32" "-ldbghelp" "-lgcc_eh" "-l:libpthread.a" "-lmsvcrt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-lmingwex" "-luser32" "-lkernel32" "-fuse-ld=lld" "-Wl,--nxcompat" "-o" "<sysroo
t>-bootstrap-tools\\release\\deps\\serde_derive-289e03abbb61a8ac.dll" "-Wl,--gc-sections" "-shared" "-Wl,--out-implib=<sysroot>-bootstrap-tools\\release\\deps\\libserde_derive-289e03abbb61a8ac.dll.a" "-nodefaultlibs" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\
rsend.o"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0librustlibx86_64-pc-windows-gnulibrsbegin.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64tmprustcIGBlzusymbols.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.00.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.01.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.02.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.03.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.04.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.05.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.06.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.07.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.08.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.09.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.10.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.11.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.12.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.13.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.14.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.serde_derive.aa02f1374bc7d472-cgu.15.rcgu.o': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.7h71bf3kbtp6y7yvtao8biynu.rcgu.rmeta': No such file or directory␍
          ld.lld: error: could not open 'C:msys64homewerustbuildx86_64-pc-windows-gnustage0-bootstrap-toolsreleasedepsserde_derive-289e03abbb61a8ac.0g3fdgiuvv97ejvcm8q5w0d3b.rcgu.o': No such file or directory␍
          ld.lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)␍
          collect2.exe: error: ld returned 1 exit status


error: could not compile `serde_derive` (lib) due to 1 previous error
Build completed unsuccessfully in 0:00:03
```

LLD cannot link correctly because the paths to object files that it sees are corrupted.

Removing `[rust] use-lld = true` from `config.toml ` switches the linker to gnu and fixes the issue.
This is a relatively recent regression (maybe in Rust, maybe in LLD package shipped by MSYS2), things worked ok 3-4 months ago.

--- a/compiler/rustc_codegen_ssa/src/back/link.rs
+++ b/compiler/rustc_codegen_ssa/src/back/link.rs
@@ -1726,8 +1726,12 @@ fn exec_linker(
         args.push_str(
             &Escape {
                 arg: arg.to_str().unwrap(),
-                // LLD also uses MSVC-like parsing for @-files by default when running on windows hosts
-                is_like_msvc: sess.target.is_like_msvc || (cfg!(windows) && flavor.uses_lld()),
+                // Windows-style escaping for @-files is used by
+                // - all linkers targeting MSVC-like targets, including LLD
+                // - all LLD flavors running on Windows hosts
+                // С/С++ compilers use Posix-style escaping (except clang-cl, which we do not use).
+                is_like_msvc: sess.target.is_like_msvc
+                    || (cfg!(windows) && flavor.uses_lld() && !flavor.uses_cc()),
             }
             .to_string(),
         );

--- a/src/bootstrap/src/core/build_steps/test.rs
+++ b/src/bootstrap/src/core/build_steps/test.rs
@@ -269,7 +269,13 @@ impl Step for Cargotest {
             .args(builder.config.test_args())
             .env("RUSTC", builder.rustc(compiler))
             .env("RUSTDOC", builder.rustdoc(compiler));
-        add_rustdoc_cargo_linker_args(&mut cmd, builder, compiler.host, LldThreads::No);
+        add_rustdoc_cargo_linker_args(
+            &mut cmd,
+            builder,
+            compiler.host,
+            LldThreads::No,
+            compiler.stage,
+        );
         cmd.delay_failure().run(builder);
     }
 }
@@ -847,7 +853,7 @@ impl Step for RustdocTheme {
             .env("CFG_RELEASE_CHANNEL", &builder.config.channel)
             .env("RUSTDOC_REAL", builder.rustdoc(self.compiler))
             .env("RUSTC_BOOTSTRAP", "1");
-        cmd.args(linker_args(builder, self.compiler.host, LldThreads::No));
+        cmd.args(linker_args(builder, self.compiler.host, LldThreads::No, self.compiler.stage));
 
         cmd.delay_failure().run(builder);
     }
@@ -1023,7 +1029,13 @@ impl Step for RustdocGUI {
         cmd.env("RUSTDOC", builder.rustdoc(self.compiler))
             .env("RUSTC", builder.rustc(self.compiler));
 
-        add_rustdoc_cargo_linker_args(&mut cmd, builder, self.compiler.host, LldThreads::No);
+        add_rustdoc_cargo_linker_args(
+            &mut cmd,
+            builder,
+            self.compiler.host,
+            LldThreads::No,
+            self.compiler.stage,
+        );
 
         for path in &builder.paths {
             if let Some(p) = helpers::is_valid_test_suite_arg(path, "tests/rustdoc-gui", builder) {
@@ -1883,7 +1895,7 @@ NOTE: if you're sure you want to do this, please open an issue as to why. In the
 
         let mut hostflags = flags.clone();
         hostflags.push(format!("-Lnative={}", builder.test_helpers_out(compiler.host).display()));
-        hostflags.extend(linker_flags(builder, compiler.host, LldThreads::No));
+        hostflags.extend(linker_flags(builder, compiler.host, LldThreads::No, compiler.stage));
 
         let mut targetflags = flags;
         targetflags.push(format!("-Lnative={}", builder.test_helpers_out(target).display()));

--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -260,7 +260,7 @@ impl Cargo {
             }
         }
 
-        for arg in linker_args(builder, compiler.host, LldThreads::Yes) {
+        for arg in linker_args(builder, compiler.host, LldThreads::Yes, 0) {
             self.hostflags.arg(&arg);
         }
 
@@ -270,10 +270,10 @@ impl Cargo {
         }
         // We want to set -Clinker using Cargo, therefore we only call `linker_flags` and not
         // `linker_args` here.
-        for flag in linker_flags(builder, target, LldThreads::Yes) {
+        for flag in linker_flags(builder, target, LldThreads::Yes, compiler.stage) {
             self.rustflags.arg(&flag);
         }
-        for arg in linker_args(builder, target, LldThreads::Yes) {
+        for arg in linker_args(builder, target, LldThreads::Yes, compiler.stage) {
             self.rustdocflags.arg(&arg);
         }
 

--- a/src/bootstrap/src/core/builder/mod.rs
+++ b/src/bootstrap/src/core/builder/mod.rs
@@ -1462,7 +1462,7 @@ impl<'a> Builder<'a> {
             cmd.arg("-Dwarnings");
         }
         cmd.arg("-Znormalize-docs");
-        cmd.args(linker_args(self, compiler.host, LldThreads::Yes));
+        cmd.args(linker_args(self, compiler.host, LldThreads::Yes, compiler.stage));
         cmd
     }
 

--- a/src/bootstrap/src/utils/helpers.rs
+++ b/src/bootstrap/src/utils/helpers.rs
@@ -430,8 +430,9 @@ pub fn linker_args(
     builder: &Builder<'_>,
     target: TargetSelection,
     lld_threads: LldThreads,
+    stage: u32,
 ) -> Vec<String> {
-    let mut args = linker_flags(builder, target, lld_threads);
+    let mut args = linker_flags(builder, target, lld_threads, stage);
 
     if let Some(linker) = builder.linker(target) {
         args.push(format!("-Clinker={}", linker.display()));
@@ -446,12 +447,18 @@ pub fn linker_flags(
     builder: &Builder<'_>,
     target: TargetSelection,
     lld_threads: LldThreads,
+    stage: u32,
 ) -> Vec<String> {
     let mut args = vec![];
     if !builder.is_lld_direct_linker(target) && builder.config.lld_mode.is_used() {
         match builder.config.lld_mode {
             LldMode::External => {
-                args.push("-Clinker-flavor=gnu-lld-cc".to_string());
+                // cfg(bootstrap) - remove after updating bootstrap compiler (#137498)
+                if stage == 0 && target.is_windows() {
+                    args.push("-Clink-arg=-fuse-ld=lld".to_string());
+                } else {
+                    args.push("-Clinker-flavor=gnu-lld-cc".to_string());
+                }
                 // FIXME(kobzol): remove this flag once MCP510 gets stabilized
                 args.push("-Zunstable-options".to_string());
             }
@@ -479,8 +486,9 @@ pub fn add_rustdoc_cargo_linker_args(
     builder: &Builder<'_>,
     target: TargetSelection,
     lld_threads: LldThreads,
+    stage: u32,
 ) {
-    let args = linker_args(builder, target, lld_threads);
+    let args = linker_args(builder, target, lld_threads, stage);
     let mut flags = cmd
         .get_envs()
         .find_map(|(k, v)| if k == OsStr::new("RUSTDOCFLAGS") { v } else { None })

